client/graphics/vertex-buffer-object.cpp

```diff
@@ -42,24 +162,88 @@ void VertexBufferObject::Update() {
 	}
 }
 
-void VertexBufferObject::Load(std::shared_ptr<MeshFile> mesh) {
-	if (!mesh) {
+void VertexBufferObject::Load(std::shared_ptr<MeshFile> meshes) {
+	auto _log = spdlog::get("console_log");
+	if (!meshes) { // someone might want to know!
+		_log->error("VertexBufferObject::Load null MeshFile");
 		return;
 	}
-	this->source_mesh = mesh;
-	std::vector<GLuint> all_indices;
-	std::vector<VertexData> all_verts;
-	GLuint vert_offset;
+	this->source_mesh = meshes;
 	this->vertex_groups.clear();
-	for (std::size_t i = 0; i < mesh->GetMeshCount(); ++i) {
-		Mesh* mfmesh = mesh->GetMesh(i);
-		vert_offset = static_cast<GLuint>(all_verts.size());
-		all_verts.insert(all_verts.end(), mfmesh->verts.begin(), mfmesh->verts.end());
+	const vertex::FormatDefinition* vertex_format = vertex::from_vftype[this->load_format];
+	if (!vertex_format) {
+		_log->error("FIXME: missing vertex::from_vftype[{}] in {}", this->load_format, __FILE__);
+		return;
+	}
+	const size_t stride = vertex_format->format_stride;
+	GLuint vert_offset{0};
+	GLuint index_offset{0};
+	uint8_t* vert_buffer{nullptr}; // bytes so we can offset by stride
+	GLuint* index_buffer{nullptr};
+	size_t all_vertices_size{0}; // the total number of vertices in the MeshFile
+	size_t all_indices_size{0}; // the total number of indices in the MeshFile
+	for (std::size_t i{0}; i < meshes->GetMeshCount(); ++i) {
+		Mesh* mfmesh = meshes->GetMesh(i);
+		all_vertices_size += mfmesh->vert_count();
+		for (ObjectGroup* objgroup : mfmesh->object_groups) {
+			all_indices_size += objgroup->indices.size();
+		}
+	}
+	if ((all_vertices_size == 0) || (all_indices_size == 0)) {
```

Comment:
Is it truly an error though to have a mesh with no data?

Reply:
the GL raises an error if you try to map or allocate buffers of zero length, so yes, I consider this an error. I also plan to refactor this part some more, so I might revisit empty meshes.